src/internal/hash.c

```diff
@@ -0,0 +1,226 @@
+ * `R_Serialize()` should be used. Version 3 is used with R >= 3.5.0, and
+ * has support for ALTREP.
+ */
+#ifdef R_VERSION
+#  if (R_VERSION >= R_Version(3, 5, 0))
+#    define USE_VERSION_3 1
+#  else
+#    define USE_VERSION_3 0
+#  endif
+#else
+#  define USE_VERSION_3 0
+#endif
+
+/*
+ * Before any R object data is serialized, `R_Serialize()` will first write out:
+ *
+ * Serialization info:
+ * - 2 bytes for `"X\n"` to declare "binary" serialization (i.e. not "ascii")
+ * - An `int` representing the serialization version
+ * - An `int` representing `R_VERSION`
+ * - An `int` representing the minimum R version where this serialization
+ *   version was supported. This is `R_Version(2,3,0)` for version 2, and
+ *   `R_Version(3,5,0)` for version 3.
+ *
+ * With version 3, it additionally writes out:
+ * - An `int` representing the `strlen()` of a `const char*` containing the
+ *   native encoding.
+ * - A `const char*` for that native encoding. The length of this comes from
+ *   the previous `int` that was written out.
+ *
+ * Since this changes between R versions, we skip these first bytes before
+ * streaming any data into the hashing algorithm.
+ *
+ * Reference to show where R appends this information:
+ * https://github.com/wch/r-source/blob/d48ecd61012fa6ae645d087d9a6e97e200c32fbc/src/main/serialize.c#L1382-L1389
+ */
+#define N_BYTES_SERIALIZATION_INFO (2 + 3 * sizeof(int))
+
+#if USE_VERSION_3
+#  define N_BYTES_N_NATIVE_ENC (sizeof(int))
+#endif
+
+// -----------------------------------------------------------------------------
+
+struct hash_state_t {
+  bool skip;
+  int n_skipped;
+#if USE_VERSION_3
+  int n_native_enc;
+#endif
+  XXH3_state_t* p_xx_state;
+};
+
+static inline struct hash_state_t new_hash_state(XXH3_state_t* p_xx_state);
+static inline int hash_version();
+static inline void hash_bytes(R_outpstream_t stream, void* p_input, int n);
+static inline void hash_char(R_outpstream_t stream, int input);
+
+sexp* rlang_hash(sexp* x) {
+  XXH3_state_t* p_xx_state = XXH3_createState();
+  XXH_errorcode err = XXH3_128bits_reset(p_xx_state);
+
+  if (err == XXH_ERROR) {
+    XXH3_freeState(p_xx_state);
+    r_abort("Couldn't initialize hash state.");
+  }
+
+  struct hash_state_t state = new_hash_state(p_xx_state);
+
+  int version = hash_version();
+
+  // Unused
+  sexp* (*hook)(sexp*, sexp*) = NULL;
+  sexp* hook_data = r_null;
+
+  struct R_outpstream_st stream;
+
+  R_InitOutPStream(
+    &stream,
+    (R_pstream_data_t) &state,
+    R_pstream_binary_format,
+    version,
+    hash_char,
+    hash_bytes,
+    hook,
+    hook_data
+  );
+
+  R_Serialize(x, &stream);
+
+  XXH128_hash_t hash = XXH3_128bits_digest(p_xx_state);
+  XXH3_freeState(p_xx_state);
+
+  // R assumes C99, so these are always defined as `uint64_t` in xxhash.h
+  XXH64_hash_t high = hash.high64;
+  XXH64_hash_t low = hash.low64;
+
+  // 32 for hash, 1 for terminating null added by `sprintf()`
+  char out[32 + 1];
```

Comment:
Overallocating has no benefit here. `PRIx64` is guaranteed to print a 64-bit integer exactly, which is guaranteed to be up to 16 hex digits. So that `sprintf` is guaranteed to write a maximum of 33 bytes.

And stack allocation is definitely faster and better for cache, especially when position independent code is involved.

Reply:
Stack allocation definitely has no measurable impact on performance here because this memory is used once per call.

I know the size is correct and it should be mentioned in the comment. I'd still overallocate as a matter of principle since that reduces risks on average, this is not about this particular case.

Reply:
If you are so worried, just use `snprintf`.

Reply:
If you're that worried, use `snprintf`.

Reply:
We are going to leave it as is